server/src/main/java/io/crate/planner/statement/CopyFromPlan.java

```diff
@@ -288,20 +298,54 @@ public static ExecutionPlan planCopyFromExecution(AnalyzedCopyFrom copyFrom,
                 lineNumberSymbol
             );
         } else {
-            sourceIndexWriterProjection = new SourceIndexWriterProjection(
+
+            List<Reference> targetColumns = table
+                .columns()
+                .stream()
+                .filter(ref ->
+                    // if targetColumns are empty CSV header/json keys are source of the truth and plan has to be adjusted in runtime to
+                    // either dynamically add columns or remove table columns, not listed in header.
+                    boundedCopyFrom.targetColumns().isEmpty()
+                        ||
+                    // Regular filtering
+                    (boundedCopyFrom.targetColumns().isEmpty() == false && boundedCopyFrom.targetColumns().contains(ref.column().name()))
+                )
+                .filter(ref -> toCollect.contains(ref) == false)
+                .collect(Collectors.toList());
+
+            toCollect.addAll(targetColumns);
```

Comment:
We unwrap `GeneratedReferences` when adding requires columns via `getSymbolsRequiredForShardIdCalc`.
Probably should do the same here?